[AArch64][GlobalISel] Take abs scalar codegen closer to SDAG

madhur13490 opened the pull request:
This patch improves codegen for scalar (<128bits) version
of llvm.abs intrinsic by using the existing non-XOR based lowering.
This takes the generated code closer to SDAG.

codegen with GISel for > 128 bit types is not very good
with these method so not doing so.

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -1021,6 +1021,11 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
       .moreElementsToNextPow2(0)
+      // Do SMax based lowering for < 128 bits.
+      .customIf([=](const LegalityQuery &Q) {
+        LLT SrcTy = Q.Types[0];
```
Should the default implementation of lower make a better guess based on legal operations?

arsenm replied:
The customIf action should also go up below the legal cases

tschuett commented:
Do you have an mca proof that smax is better than xor?

madhur13490 commented:
> Do you have an mca proof that smax is better than xor?

I understand the question but my intention of the patch is to take the codegen closer to what SDAG is doing. However, if you see the i8 case, it is not quite close yet and I am also not sure if the final assembly is right. Ideally, it should just be cmp+cneg (which is what SDAG) is doing.

If we get it closer to SDAG's codegen, then the benefit would be visible as Global ISel's current lowering leads to 3 instructions whereas SDAG is using just cmp+cneg.  
I am looking for suggestions around this.

aemerson commented:
I agree with Matt, `lower()` should be checking for the legality of SMAX and use that if available, falling back to the current XOR method, see `TargetLowering::expandABS()`. But for AArch64 that's only legal with the CSSC extension, so you'll need to check why we generate the compare lowering in DAG at all.

madhur13490 committed:
[AArch64][GlobalISel] Take abs scalar codegen closer to SDAG

This patch improves codegen for scalar (<128bits) version
of llvm.abs intrinsic by using the existing non-XOR based lowering.
This takes the generated code closer to SDAG.

codegen with GISel for > 128 bit types is not very good
with these method so not doing so.

davemgreen commented:
Given we have a csneg instruction, do we just want `cmp x, 0; csneg x, x`? i.e I would expect us to want to lower to a compare+neg+select (which isn't either of the existing lowerAbs* methods, and might not be easy to tell generically that we want to generate code for, if there is not an obvious way to check for csneg).

madhur13490 commented:
> Given we have a csneg instruction, do we just want `cmp x, 0; csneg x, x`? i.e I would expect us to want to lower to a compare+neg+select (which isn't either of the existing lowerAbs* methods, and might not be easy to tell generically that we want to generate code for, if there is not an obvious way to check for csneg).

@davemgreen Yes, I was experimenting with this idea locally after this patch but was running into legalizing G_SELECT issue. I have fixed it now and the new commit now introduces cneg-based lowering. Please have a look.

@arsenm @aemerson Given that there is a different approach now, do you still think `lower()` should be modified? If so, how do we reliably detect? Can you please elaborate a bit more?

arsenm commented:
> @arsenm @aemerson Given that there is a different approach now, do you still think `lower()` should be modified? If so, how do we reliably detect? Can you please elaborate a bit more?

Yes. Currently the lower implementation just calls lowerAbsToAddXor. You can conditionally call your new implementation based on isLegalOrCustom for the type with the relevant opcodes used in the expansion

tschuett commented:
Even if SMAX and lowerAbsToAddXor are legal, shouldn't we have the means to say for FOO cores, we prefer SMAX, and for BAR cores we prefer Xor?

madhur13490 commented:
> > @arsenm @aemerson Given that there is a different approach now, do you still think `lower()` should be modified? If so, how do we reliably detect? Can you please elaborate a bit more?
> 
> Yes. Currently the lower implementation just calls lowerAbsToAddXor. You can conditionally call your new implementation based on isLegalOrCustom for the type with the relevant opcodes used in the expansion

I have updated. Please have a look.

davemgreen commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -515,6 +515,7 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
   getActionDefinitionsBuilder(G_ICMP)
       .legalFor({{s32, s32},
                  {s32, s64},
+                 {s64, s64},
```
I think this and the select below is marking a s64 condition code as legal, which Im not sure is what we would want. We have the clampScalar(0, s32, s32) below.

madhur13490 replied:
Well, s64 case doesn't work without this. Why don't want this to be legal?

madhur13490 replied:
{s64, s64} is legal for G_SELECT but further the index is clamped to s32. I don't understand why this is done. 
Inspiring from this I made {s64, s64} legal for G_ICMP.

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -3914,8 +3914,12 @@ LegalizerHelper::lower(MachineInstr &MI, unsigned TypeIdx, LLT LowerHintTy) {
-  case G_ABS:
+  case G_ABS: {
+    LLT Ty = MRI.getType(MI.getOperand(0).getReg());
+    if (LI.isLegalOrCustom({G_ABS, Ty}))
```
Checking the G_ABS legality doesn't make sense in the G_ABS lowering implementation. You should need to check the operations in the expansion

davemgreen replied:
I think the problem might be that there is no good way with the isLegal calls to check if there is a single instruction that does "select and conditional neg" at the same time, in the way csneg would do.

arsenm replied:
For reference, the DAG has this chain of implementations: https://github.com/llvm/llvm-project/blob/8f68022f8e6e54d1aeae4ed301f5a015963089b7/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L8962

madhur13490 replied:
Checking for G_SELECT and G_ICMP leads to assertion when compiled for  AMDGPU in abs.ll.
`const T &llvm::ArrayRef<llvm::LLT>::operator[](size_t) const [T = llvm::LLT]: Assertion `Index < Length && "Invalid index!"' failed.`

madhur13490 replied:
```
LLT Ty = MRI.getType(MI.getOperand(0).getReg());
dbgs() << LI.isLegalOrCustom({G_ICMP, Ty});
```
Leads to an assert for llvm.abs.ll test. This doesn't need my patch to trigger an assert. 
@arsenm Thoughts? 
If this is uncovering a latent issue then I would like to go with legalizeCustom() and follow @davemgreen's suggestion. This also aligns with what RISCV is doing.

arsenm replied:
Checking for them how? 

The specific instruction AArch64 wants to use is only indirectly related

madhur13490 replied:
Found that the assert is triggering for `llvm.abs.i64` case in `AMDGPU/GlobalISel/llvm.abs.ll`. Constraining the lowering to < 8 bytes has no assert. Do you know what might be causing this?

arsenm replied:
Don't know what that has to do with abs, but this isn't a well formed query for G_ICMP legality. G_ICMP has 2 type indexes

madhur13490 replied:
Hmm... I was looking at how G_SSUBSAT is using G_SMIN and followed the same approach. 
Anyway, updated the branch to check both type indexes.

davemgreen reviewed (commented):
I think considering these are fairly AArch64 specific instructions, it might be best to just make this custom, in the same way we have it for SDAG.

madhur13490 commented:
> I think considering these are fairly AArch64 specific instructions, it might be best to just make this custom, in the same way we have it for SDAG.

Do you mean just use `customifScalar` and not have anything in `lower()`?

madhur13490 commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -8219,6 +8223,19 @@ LegalizerHelper::lowerAbsToMaxNeg(MachineInstr &MI) {
+  Register DestReg = MI.getOperand(0).getReg();
+  LLT Ty = MRI.getType(SrcReg);
+  auto Zero = MIRBuilder.buildConstant(Ty, 0).getReg(0);
+  auto Sub = MIRBuilder.buildSub(Ty, Zero, SrcReg).getReg(0);
```
Done. Thanks!

davemgreen commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -3914,8 +3914,12 @@ LegalizerHelper::lower(MachineInstr &MI, unsigned TypeIdx, LLT LowerHintTy) {
+  case G_ABS: {
+    LLT Ty = MRI.getType(MI.getOperand(0).getReg());
+    if (Ty.isScalar() &&  LI.isLegalOrCustom({G_ICMP, {Ty, Ty}}))
+        return lowerAbsToCNeg(MI);
```
I personally don't know if there is a good check that captures what we would want through the isLegalOrCustom interface. I think many targets would want the ashr/AddXor form even if they have legal sub+sel. I would remove this code and just keep the AArch64LegalizerInfo::legalizeCustom method.

arsenm replied:
Reusing the same type for the compare result isn't right, but we don't have a great API for this. We kind of want the equivalent of getSetCCResultType. A more awkward solution would be to query with s1 and then look through the legalize actions until you find transforms on the original. Failing that, using i1 as the basis of the query is a more reasonable choice I think

tschuett replied:
My main point is that we should do something akin to the DAG. We check for the legality of icmp and then select a select without checking whether it is legal or not. The icmp check is insufficient.

madhur13490 replied:
@tschuett How would that handle @arsenm's concern?

I agree with @davemgreen's approach and will update the patch accordingly. I would not like to touch `lower` as of now and handle it in `legalizeCustom`. Given that we don't have the right APIs, the right approach would be handle this separately.

tschuett replied:
These are independent issues. We have no API to query the result type from compares. But building a Select without checking whether it is legal to do so may cause trouble.

madhur13490 replied:
but that would still reuse the same type (as @arsenm pointed out). Isn't it? So, how would it solve the core issue?
I understand the lack of API and thus I would like to avoid changing `lower()` in this patch.

arsenm commented:
> Do you mean just use `customifScalar` and not have anything in `lower()`?

Regardless of what AArch64 wants or does, we should still try to pick a preferred lowering in the default implementation of lower.

davemgreen reviewed (approved):
This LGTM. Thanks.
